cli/cmd/uninstall/uninstall.go

```diff
@@ -361,10 +427,106 @@ func (c *Command) uninstallHelmRelease(releaseName, namespace, releaseType strin
 		c.UI.Output("Uninstall result: %s", res.Info, terminal.WithInfoStyle())
 		return nil
 	}
+
 	c.UI.Output(fmt.Sprintf("Successfully uninstalled %s Helm release.", releaseType), terminal.WithSuccessStyle())
 	return nil
 }
 
+// fetchCustomResources gets a list of all custom resources deployed in the
+// cluster that are managed by Consul.
+func (c *Command) fetchCustomResources() ([]unstructured.Unstructured, error) {
+	crs := make([]unstructured.Unstructured, 0)
+
+	crds, err := c.apiext.ApiextensionsV1().CustomResourceDefinitions().List(c.Ctx, metav1.ListOptions{
+		LabelSelector: "app=consul",
+	})
+	if err != nil {
+		return crs, err
+	}
+
+	for _, crd := range crds.Items {
+		for _, version := range crd.Spec.Versions {
+			target := schema.GroupVersionResource{
+				Group:    crd.Spec.Group,
+				Version:  version.Name,
+				Resource: crd.Spec.Names.Plural,
+			}
+
+			crList, err := c.dynamic.Resource(target).List(c.Ctx, metav1.ListOptions{})
+			if common.IgnoreNotFoundError(err) != nil {
+				return nil, err
+			}
+			if crList != nil {
+				crs = append(crs, crList.Items...)
+			}
+		}
+	}
+
+	return crs, nil
+}
+
+// deleteCustomResources takes a list of unstructured custom resources and
+// sends a request to each one to be deleted.
+func (c *Command) deleteCustomResources(crs []unstructured.Unstructured) error {
+	for _, cr := range crs {
+		apiVersion := strings.Split(cr.GetAPIVersion(), "/")
+		group, version := apiVersion[0], apiVersion[1]
+		if group == "" || version == "" {
+			return fmt.Errorf("malformed api version: %s", apiVersion)
+		}
+
+		target := schema.GroupVersionResource{
+			Group:    group,
+			Version:  version,
+			Resource: strings.ToLower(cr.GetKind()),
+		}
+
+		err := c.dynamic.
+			Resource(target).
+			Namespace(cr.GetNamespace()).
+			Delete(c.Ctx, cr.GetName(), metav1.DeleteOptions{})
+		if common.IgnoreNotFoundError(err) != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// patchCustomResources takes a list of unstructured custom resources and
+// sends a request to each one to patch its finalizers to an empty list.
+func (c *Command) patchCustomResources(crs []unstructured.Unstructured) error {
+	finalizerPatch := []byte(`[{
+		"op": "replace",
+		"path": "/metadata/finalizers",
+		"value": []
+	}]`)
+
+	for _, cr := range crs {
+		apiVersion := strings.Split(cr.GetAPIVersion(), "/")
+		group, version := apiVersion[0], apiVersion[1]
+		if group == "" || version == "" {
+			return fmt.Errorf("malformed api version: %s", apiVersion)
+		}
+
+		target := schema.GroupVersionResource{
+			Group:    group,
+			Version:  version,
+			Resource: strings.ToLower(cr.GetKind()),
+		}
+
+		_, err := c.dynamic.
+			Resource(target).
+			Namespace(cr.GetNamespace()).
+			Patch(c.Ctx, cr.GetName(), types.JSONPatchType, finalizerPatch, metav1.PatchOptions{})
+		if common.IgnoreNotFoundError(err) != nil {
```

Comment:
Similar to the other ignore case, we will get here after 5 deletion attempts have failed to remove all custom resources. This can mean that there is a deadlock that makes the deletion impossible, or that the deletion is just taking a long time to complete. In the latter case, the deletion request may succeed before the finalizer patch request is sent, but after this function starts executing. So we may be trying to patch a custom resource that no longer exists and don't want to fail the uninstall task.

Reply:
You could use `k8serrors.IsNotFound` instead

Reply:
Oh nice! I didn't know about that!

Reply:
I'm now using the `k8serrors.IsNotFound` instead and deleted the solution I wrote.